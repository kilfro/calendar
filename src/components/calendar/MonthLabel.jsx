import React from 'react'
import { connect } from 'react-redux'
import '@style/calendar/monthLabel.less'

const MonthLabel = ({ month }) => {
    return (
        <div className='month-label'>
            {month.toLocaleDateString('ru-RU', { month: 'long', year: 'numeric' }).replace(' Ð³.', '')}
        </div>
    )
}

const mapStateToProps = state => {
    return {
        month: state.month
    }
}

export default connect(mapStateToProps)(MonthLabel)