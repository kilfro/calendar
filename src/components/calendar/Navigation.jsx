import React from 'react'
import { connect } from 'react-redux'
import { ChevronLeft, ChevronRight } from 'react-bootstrap-icons'
import Button from '../Button'
import { prevMonth, nextMonth, goToToday } from '@store/actions'
import '@style/calendar/navigation.less'

export const Navigation = ({ month, prevMonth, nextMonth, goToToday }) => {
    return (
        <div className='navigation'>
            <Button clickHandler={() => prevMonth(month)}><ChevronLeft /></Button>
            <Button clickHandler={() => goToToday()} filled={true}>Сегодня</Button>
            <Button clickHandler={() => nextMonth(month)}><ChevronRight /></Button>
        </div>
    )
}

const mapStateToProps = state => {
    return {
        month: state.month
    }
}

const mapDispatchToProps = {
    prevMonth,
    nextMonth,
    goToToday
}

export default connect(mapStateToProps, mapDispatchToProps)(Navigation)
