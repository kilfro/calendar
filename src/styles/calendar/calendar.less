.calendar {
  -moz-user-select: none;
  -khtml-user-select: none;
  user-select: none;
  width: fit-content;
  margin: 1rem;
}
