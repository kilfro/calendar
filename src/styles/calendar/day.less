@accent-color: black;

.day {
  border-radius: 50%;
  height: 2rem;
  width: 2rem;
  margin: 0.5rem;
  font-size: 1.3rem;
  font-weight: bold;
  display: flex;
  align-items: center;
  justify-content: center;
  color: @accent-color;
  cursor: pointer;

  &.has-task::after {
    content: '.';
    position: absolute;
    margin-top: 1rem;
    font-size: 2rem;
  }

  &.weekend {
    color: #999;
  }

  &.selected {
    background-color: @accent-color;
    color: white;
    border: none !important;
  }

  &.today {
    @accent-color: red;
    color: @accent-color;
  }

  &.today.selected {
    @accent-color: red;
    background-color: @accent-color;
    color: white;
    border: none !important;
  }

  &.label {
    cursor: auto;
    font-size: 1rem;
    font-weight: initial;
  }
}
