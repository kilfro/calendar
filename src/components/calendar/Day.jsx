import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { selectDate } from '@store/actions'
import '@style/calendar/day.less'

const Day = ({ date, now, selected, hasTask, selectDate }) => {

    const isToday = date.toDateString() === now.toDateString()
    const isSelected = date.toDateString() === selected.toDateString()

    const dayStyle = () => {
        let classes = ['day']

        if (isSelected) {
            classes.push('selected')
        }

        if (isToday) {
            classes.push('today')
        }

        if ([5, 6].includes(date.getWeekDay())) {
            classes.push('weekend')
        }

        if (hasTask) {
            classes.push('has-task')
        }

        return classes.join(' ')
    }

    return (
        <div className={dayStyle()} onClick={() => selectDate(date)}>
            {date.getDate()}
        </div>
    )
}

Day.propTypes = {
    date: PropTypes.object.isRequired,
}

Day.defaultProps = {
    hasTask: false
}

const mapStateToProps = state => {
    return {
        now: state.now,
        selected: state.selected
    }
}

const mapDispatchToProps = {
    selectDate
}

export default connect(mapStateToProps, mapDispatchToProps)(Day)
