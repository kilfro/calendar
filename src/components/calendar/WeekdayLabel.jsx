import React from 'react'
import '@style/calendar/day.less'
import '@style/calendar/week.less'

const WeekdayLabel = () => {
    const labels = ['ПН', 'ВТ', 'СР', 'ЧТ', 'ПТ', 'СБ', 'ВС']
    return (
        <div className='week label'>
            {labels.map(lab =>
                <div className='day label' key={lab}>{lab}</div>
            )}
        </div>
    )
}

export default WeekdayLabel