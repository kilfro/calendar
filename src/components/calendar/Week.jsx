import React from 'react'
import PropTypes from 'prop-types'
import Day from './Day'
import '@style/calendar/week.less'

const Week = ({ days }) =>
    <div className='week'>
        {days.map(day => <Day date={day} key={day.getTime()} />)}
    </div>

Week.propTypes = {
    days: PropTypes.arrayOf(PropTypes.object).isRequired
}

export default Week