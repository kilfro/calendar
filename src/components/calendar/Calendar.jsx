import '@style/calendar/calendar.less'

import React, { useEffect } from 'react'

import Month from './Month'
import MonthLabel from './MonthLabel'
import Navigation from './Navigation'
import WeekdayLabel from './WeekdayLabel'
import { connect } from 'react-redux'
import { goToToday } from '../../store/actions'

export const Calendar = ({ goToToday }) => {
    useEffect(() => { goToToday() }, [])

    return (
        <div className='calendar'>
            <Navigation />
            <MonthLabel />
            <WeekdayLabel />
            <Month />
        </div>
    )
}

const mapDispatchToProps = {
    goToToday
}

export default connect(null, mapDispatchToProps)(Calendar)