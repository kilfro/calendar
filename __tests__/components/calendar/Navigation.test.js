import React from 'react'
import { shallow, mount } from 'enzyme'
import { Navigation } from '../../../src/components/calendar/Navigation'
import { ChevronLeft, ChevronRight } from 'react-bootstrap-icons'

describe('Navigation component', () => {
    test('should renders correctly', () => {
        const component = shallow(<Navigation />)
        
        expect(component).toMatchSnapshot()
    })

    test('should call prevMonth function', () => {
        const month = new Date()
        const prevFunc = jest.fn()
        const component = mount(
            <Navigation month={month} prevMonth={prevFunc} />
        )

        component.find(ChevronLeft).simulate('click')

        expect(prevFunc).toHaveBeenCalled()
        expect(prevFunc).toHaveBeenCalledWith(month)
    })

    test('should call nextMonth function', () => {
        const month = new Date()
        const nextFunc = jest.fn()
        const component = mount(
            <Navigation month={month} nextMonth={nextFunc} />
        )

        component.find(ChevronRight).simulate('click')

        expect(nextFunc).toHaveBeenCalled()
        expect(nextFunc).toHaveBeenCalledWith(month)
    })

    test('should call goToToday function', () => {
        const goToFunction = jest.fn()
        const component = mount(
            <Navigation goToToday={goToFunction} />
        )

        component.find('div[children="Сегодня"]').simulate('click')

        expect(goToFunction).toHaveBeenCalled()
    })
})
