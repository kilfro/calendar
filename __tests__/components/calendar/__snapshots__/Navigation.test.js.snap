// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation component should renders correctly 1`] = `
<div
  className="navigation"
>
  <Button
    clickHandler={[Function]}
    filled={false}
  >
    <ForwardRef
      color="currentColor"
      size="1em"
    />
  </Button>
  <Button
    clickHandler={[Function]}
    filled={true}
  >
    Сегодня
  </Button>
  <Button
    clickHandler={[Function]}
    filled={false}
  >
    <ForwardRef
      color="currentColor"
      size="1em"
    />
  </Button>
</div>
`;
